cpmpy/solvers/pindakaas.py

```diff
@@ -0,0 +1,286 @@
+            if has_sol:
+                # fill in variable values
+                for cpm_var in self.user_vars:
+                    if cpm_var.name in self._varmap:
+                        lit = self.solver_var(cpm_var)
+                        cpm_var._value = result.value(lit)
+                        if cpm_var._value is None:
+                            cpm_var._value = True  # dummy value
+                    else:  # if pindakaas does not know the literal, it will error
+                        cpm_var._value = None
+                self.core = None
+            else:  # clear values of variables
+                for cpm_var in self.user_vars:
+                    cpm_var._value = None
+                # we have to save the unsat core here, as the result object does not live beyond this solve call
+                if assumptions is not None:
+                    self.core = [
+                        x
+                        for x, s_x in zip(assumptions, solver_assumptions)
+                        if result.failed(s_x)
+                    ]
+
+        return has_sol
+
+    def solver_var(self, cpm_var):
+        if isinstance(cpm_var, NegBoolView):  # negative literal
+            # get inner variable and return its negated solver var
+            return ~self.solver_var(cpm_var._bv)
+        elif isinstance(cpm_var, _BoolVarImpl):  # positive literal
+            # insert if new
+            if cpm_var.name not in self._varmap:
+                (self._varmap[cpm_var.name],) = self.pdk_solver.new_vars(1)
+            return self._varmap[cpm_var.name]
+        else:
+            raise NotImplementedError(
+                f"{self.name}: unexpected variable {cpm_var} of type {type(cpm_var)} not supported"
+            )
+
+    def transform(self, cpm_expr):
+        cpm_cons = toplevel_list(cpm_expr)
+        cpm_cons = no_partial_functions(cpm_cons)
+        cpm_cons = decompose_in_tree(cpm_cons)
+        cpm_cons = simplify_boolean(cpm_cons)
+        cpm_cons = flatten_constraint(cpm_cons)  # flat normal form
+        cpm_cons = only_bv_reifies(cpm_cons)
+        cpm_cons = only_implies(cpm_cons)
+        cpm_cons = linearize_constraint(
+            cpm_cons, supported=frozenset({"sum", "wsum", "and", "or"})
+        )
+        return cpm_cons
+
+    def add(self, cpm_expr_orig):
+        import pindakaas as pdk
+
+        if self.unsatisfiable:
+            return self
+
+        # add new user vars to the set
+        get_variables(cpm_expr_orig, collect=self.user_vars)
+
+        # transform and post the constraints
+        try:
+            for cpm_expr in self.transform(cpm_expr_orig):
+                self._add_transformed(cpm_expr)
+        except pdk.Unsatisfiable:
+            self.unsatisfiable = True
+
+        return self
+
+    __add__ = add  # avoid redirect in superclass
+
+    def _add_transformed(self, cpm_expr, conditions=[]):
+        """Add for a single, *transformed* expression, implied by conditions."""
```

Comment:
I was confused at first what this function did, or why it was required at all (instead of just doing it in `add`).
But now I get that it is because of the implications.
Might be better to rename this to `_post_constraint`, which is also how it's called in the OR-Tools interface.